### PR TITLE
Introduce Que benchmark, add extra metrics, fix performance issue

### PR DESCRIPTION
**Pull request by lawrencejones**
Due to a bug introduced in the cursor patch, Que would trigger the wake_interval sleep more often than was appropriate. This PR ensures that when Que is working at capacity, we see approximately 90% of worker utilisation. 

We also add a benchmarking suite that can help with debugging these issues in future.

---

## Que benchmark

Create a benchmark that can be used to test changes to Que in a
reliable, repeatable fashion. This leverages kubernetes to run several
replicas of sample Que workers, with a script to publish large numbers
of jobs to the que_jobs table with randomised priorities.

We make a small modification to fix an incorrect constant in the bin/que
script. This wasn't noticed before because we always set the parameter
and avoided execution of the bad code path.

## Add metrics and improve work loop

We had a performance bug where introducing cursors could cause the Que
worker to trigger the wake_interval sleep when there were jobs remaining
in the queue to be worked. This is because we'd attempt to lock a job
with a job_id > cursor, find no job that matched that condition, and
return to the Que worker that there are no jobs.

This is incorrect, as there are many jobs in the queue, they just didn't
match our initial cursored query. We should instead immediately retry
acquiring a job without the cursor to ensure we don't leave jobs
unworked in the queue.

There was also another bug that was triggered less frequently where we
would return nil when we hit the MVCC race condition. We'd also return
nil here, which would be wrong as it signals there are no jobs left,
when actually we've just raced against another worker. Fixing this
improved throughput by about 10%.

To debug these issues we introduce a variety of new metrics, which
include the following:

  que_worker_sleeping_total
  que_worker_sleeping_seconds_total
  que_worker_job_exists__total
  que_worker_job_exists_seconds_total
  que_worker_job_unlock_total
  que_worker_job_unlock_seconds_total

This allows further breakdown of Que worker behaviour to ensure we cover
the significant factors of Que performance, and can detect when we're
sleeping on the job!

### CODE REVIEW

File: lib/que/locker.rb [hunk 1]
```diff
@@ -24,9 +24,23 @@ def initialize(queue:, cursor_expiry:, metrics:)
...
+      job = lock_job
+
+      # Becuase we were using a cursor when we tried to lock this job, if we fail to find
+      # a job it is not necessarily the case that there aren't jobs in the queue. We may
+      # have been excluding candidate jobs due to the cursor that are now due to be
+      # worked.
+      #
+      # We should attempt to lock again after resetting our cursor to make sure we include
+      # jobs that were excluded in the first attempt.
+      if job.nil? && @cursor
```
Comment (dmagliola) [inline lib/que/locker.rb]: @cursor is 0 if there is no cursor, right?
If that's the case, @cursor is always true, and we'll go into this if no matter what.

So, this is either a bug, or we're just redundantly checking `@cursor`, and we'll reset and reload anyway even if there was no cursor.

File: lib/que/locker.rb [hunk 2]
```diff
@@ -24,9 +24,23 @@ def initialize(queue:, cursor_expiry:, metrics:)
     def with_locked_job
       reset_cursor if cursor_expired?
 
+      job = lock_job
+
+      # Becuase we were using a cursor when we tried to lock this job, if we fail to find
```
Comment (dmagliola) [inline lib/que/locker.rb]: Bec**au**se (minor annoyance, sorry)

File: lib/que/locker.rb [hunk 3]
```diff
@@ -24,9 +24,23 @@ def initialize(queue:, cursor_expiry:, metrics:)
...
+      # jobs that were excluded in the first attempt.
+      if job.nil? && @cursor
+        reset_cursor
+        job = lock_job
+      end
+
       # Check that the job hasn't just been worked by another worker (it's possible to
       # lock a job that's just been destroyed because pg locks don't obey MVCC). If it has
-      # been worked, act as if we've worked it.
+      # been worked, return nil to immediately retry locking.
```
Comment (dmagliola) [inline lib/que/locker.rb]: I think it'd be useful to clarify in a comment at the top of this function that "returning nil" will get this method called again immediately, while "yielding nil" makes the worker sleep a bit and then call this method again.

The distinction is a bit subtle, and this particular idea commented here is not that obvious (at least to me, I had to go look at `worker.rb` to make sure I was understanding this correctly)

File: Dockerfile.benchmark [hunk 4]
```diff
@@ -0,0 +1,6 @@
+FROM ruby:2.5
```
Comment (dmagliola) [inline Dockerfile.benchmark]: Just a note: Our `.ruby-version` for this project is `2.4.1`, while the dockerfile for the benchmark specifies `2.5`
I'm assuming this is not a big issue, and this is probably best solved in a separate PR if the solution is to update `.rubyversion`, but i think it'd be good to make these two consistent.

File: benchmark/README.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+```shell
+$ docker build -t gocardless/que-benchmark -f Dockerfile.benchmark .
+$ docker push gocardless/que-benchmark
+```
+
+Once pushed, load the Kubernetes resources into your cluster:
+
+```shell
+$ kubectl create namespace que-benchmarking
+$ kubectl --namespace apply -f benchmark/deployment.yaml
```
Comment (dmagliola) [inline benchmark/README.md]: `kubectl --namespace=que-benchmarking apply -f benchmark/deployment.yaml`

File: benchmark/README.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+deployment "que-postgres" configured
+configmap "que-postgres-init-scripts" configured
+service "que-postgres" configured
+deployment "que-benchmark" configured
+```
+
+Once all pods are ready, you should be able to see them using kubectl:
+
+```
+$ kubectl --namespace que-benchmark get pods
```
Comment (dmagliola) [inline benchmark/README.md]: `kubectl --namespace que-benchmarking get pods`

File: benchmark/README.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+que-benchmark-79c779b6db-ntqv9   1/1       Running   0          43m
+que-benchmark-79c779b6db-r8gkt   1/1       Running   0          43m
+que-benchmark-79c779b6db-z7jm4   1/1       Running   0          43m
+que-postgres-5c5d69949d-rjpl6    1/1       Running   0          4h
+```
+
+You can now exec into one of the que containers to kickoff a benchmark:
+
+```
+$ kubectl --namespace que-benchmark exec -i que-benchmark-79c779b6db-2mk9q bash
```
Comment (dmagliola) [inline benchmark/README.md]: `kubectl --namespace que-benchmarking`

**Comment by dmagliola** [comment]
10/10 - Confirmed it works:

screen shot 2018-05-02 at 17 26 04

- Benchmark code works beautifully, and it's super well encapsulated such that anyone can easily run this benchmark.
- New version of the code processes more jobs / s (top graph), with higher utilization (bottom graph) and gets through the 10,000 jobs *a lot* faster than old version (width of the red rectangles).

**Comment by gotjosh** [comment]
Super excited to see this go through! Top work @lawrencejones  🎖

File: lib/que/locker.rb (hunk 2 above)
Comment (lawrencejones) [reply lib/que/locker.rb]: No problem- good spot!

File: Dockerfile.benchmark (hunk 4 above)
Comment (lawrencejones) [reply Dockerfile.benchmark]: Yep, worth bumping them to match.

**Pull request merged**